### ISSUE TITLE
minimize_constrained interface

### ISSUE THREAD

**Issue by antonior92** [scipy/scipy#7648]
Hi all!

I am working on a constrained optimizer solver for my GSoC project. I started on a separate repository and I am working now on a pull request to integrate it to SciPy. In order to prooced we have to make the following decision: should we fit the constrained solver to the ``minimize`` interface or should we create a new one?

The best idea we came up with to try to fit the new solver into the ``minimize`` interface is to create a ``Constraints`` class and use it to pass the constraints specifications.

I am, however, more and more conviced that the best option would be to create a new function ``minimized_constrained``. This new interface would be used to constrained minimization problems and I would include both SLSQP and COBLYA to it. The ``minimize`` interface would be reserved for unconstrained and bound-constrained problems.

I think it would be better for the following reasons:

1. The ``minimize`` interface would be kept simple and reserved for unconstrained problems and problems with simple bound constraints.
2. We would have more freedom designing a ``minimize_constrained`` interface that best suit our needs.
3. The burden of implementing and including SLSQP and COBLYA to a new interface is not much greater than the burden of fitting the constrained solver being implemented into minimize.
4. This new interface could be used for new constrained solvers included to scipy.

Let me elaborate on 1: I want to include finite-derivative options, quasi-newton options, different ways of passing constraints, different ways of passing second order information... All of that requires a great deal of documentation and the need of including several arguments to the function signature. And my point is: why to complicate ``minimize`` interface for something that will only be used by a single solver method?

The idea is to create a new interface so we keep ``minimize`` interface clean and easy to use. It would also give us more freedom for creating a better ``minimize_constrained`` interface.

I would much appreciate any feedback on this.

**Comment by pv** [comment]
From this description it is not clear what in your suggested new 
interface differs substantially from that in minimize(). What are the 
new arguments that are crucial to include to the interface?

**Comment by pv** [comment]
Ok, thanks, I see now there's a long discussion about this in the ticket 
in the other repository.

**Comment by antonior92** [comment]
What would have to be included to the minimize interface:
- Second order information
    - ``hess_lagr``
    - ``hess_ineq``
    - ``hess_eq``
- Options for quasi-newton options:
   - ``BFGS``
   - ``SR1``
   - ``L-BFGS``
- Options for finite differentiation.
- The current way of passing constraints in minimize is not very powerful. So I would like to introduce two alternative interfaces. One matlab-like interface:
```
minimize fun(x)
subject to: constr_ineq(x) <= 0
            constr_eq(x) = 0
            A_ineq x <= b_ineq
            A_eq x = b_eq
            lb <= x <= ub
```
and one interface similar to the one used by CUTEst, IPOPT, JuMP,...:
```
minimize fun(x)
subject to: c_lb <= constr(x) <= c_ub
            lb <= x <= ub
```
which specifies equality constraints by setting c_lb = c_ub and specifies linear constraints using an auxiliar vector.
- I would also want to include explicit support for sparse matrix and LinearOperator.
- Better callback functions and stop criteria options would also be easier to set in a new interface.

Anyway, my point is: this is a little to much to include in the interface just to be used for a single method. It would only make things more complicated. Why not just creating two clean separate interfaces?

**Comment by nmayorov** [comment]
Hey, Antonio. I have some concerns here. 

If we consider `minimize_constrained` as a generic interface to constrained optimization, then I'm not sure that everything you have listed belongs to it. That is we should include explicitly only such arguments that appear in every solver. For example, options for quasi-Newton approximation are not from such category in my opinion. 

So I guess in the end it will look similar to `minimize` or [`solve_ivp`](http://scipy.github.io/devdocs/generated/scipy.integrate.solve_ivp.html#scipy.integrate.solve_ivp). 

If we consider this, how can we make `minimize_constrained` more convenient than `minimize`? What are you thoughts?

-------------

> The current way of passing constraints in minimize is not very powerful. So I would like to introduce two alternative interfaces.

How do you plan to achieve that? Perhaps `Constraints` class may have a place here?

**Comment by pv** [comment]
As I see it the broad intent of minimize() had two aims (i) provide a 
single "domain-specific language" for specifying the optimization 
problem, and (ii) unify the naming conventions of several solver tuning 
options such as termination tolerances insofar as it's possible.
.
An optimization problem DSL is something that many optimization packages 
have. Now, one concern in adding a new entry point is that then you have 
two incompatible ways of specifying constrained problems.
.
Of course, it may be the minimize() way of problem specification is not 
so convenient. But if this is the problem, maybe there is some way to 
address it?

**Comment by antonior92** [comment]
You guys are right. I am giving up too easily on ``minimize``. And actually most of derivative options I intend to implement (finite difference and quasi-Newton options) could be used in other solvers as well.

Maybe we could extend the constraints definition in the following way:
```
    constraints : dict or sequence of dict
        Constraints definition.
        Each constraint is defined in a dictionary with fields:
            type : str
                Constraint type: 
                - 'eq' for equality constraint of the type:
                     fun(x) = 0
                - 'ineq' for inequality constraint of the type:
                     fun(x) >= 0
                - 'interval' for a constraint of the type:
                     lb <= fun(x) <= ub
            fun : callable
                The function defining the constraint.
                      fun(x) -> array_like, shape (m,)
            jac : {‘2-point’, ‘3-point’, ‘cs’, callable}, optional
                If a callable is passed, it should be the Jacobian matrix of `fun`:
                     jac(x) -> ndarray (or sparse matrix), shape (m, n)
                Otherwise, ‘2-point’, ‘3-point’, ‘cs’ specify the type of finite difference approximation
                used.
            jac_sparsity : {None, array_like, sparse matrix}, optional
                  Defines the sparsity structure of the Jacobian matrix for finite difference estimation, its 
                  shape must be (m, n).
            hess : {‘2-point’, ‘3-point’, callable}, optional
                 If a callable is passed, the Hessian matrix of `dot(fun, v)`:
                     hess(x, v) -> ndarray (or sparse matrix or LinearOperator), shape (n, n)
                Otherwise, ‘2-point’, ‘3-point’ specify the type of finite difference approximation
                used.
            lb : array_like, shape (m, ), optional
                 Lower bound for interval constraint;
            ub :  array_like, shape (m, ), optional
                  Upper bound for interval constraint;
            linear : flag, optional
                 Flag indicating if the constraint is linear
            args : sequence, optional
                Extra arguments to be passed to `fun`, `jac`, `hess`.
```

**Comment by nmayorov** [comment]
OK, but wouldn't it be better to create a separate class for this? 

That is we will have the old format (with dicts, possible to be deprecated) and the new format --- a single class, which will a) provide convenient ways to specify constraints in one of the formats you mentioned and b) make adaptations / conversions inside to support different routines (old and new).

**Comment by antonior92** [comment]
I see some advantages using a sequence of dictionaries as I have specified above:
1. It is easy to pass matlab-like constraints
```
           constr_ineq(x) <= 0
            constr_eq(x) = 0
            A_ineq x <= b_ineq
            A_eq x = b_eq
            lb <= x <= ub
```
2. It is easy to pass constraints similar to the ones used by CUTEst, IPOPT, JuMP. (Both under the same interface)
```
            c_lb <= constr(x) <= c_ub
            lb <= x <= ub
```
3. It is backward compatible, so we will not have to explain several different ways of passing constraints (what would make the documentation much more confusing)

**Comment by nmayorov** [comment]
Uhm, I don't know. I think the same can be achieved with a class. And using dictionaries doesn't fit into scipy/numpy style of things.

**Comment by antonior92** [comment]
I actually does not opose the idea of using Constraint classes. But I actually kind of like the idea of passing the constraints as a sequence of specification. This make possible to combine different interfaces seamlesly.

**Comment by antonior92** [comment]
And I think it is better to have sequence of specifications than different interfaces for matlab-like constraints:
```
minimize fun(x)
subject to: constr_ineq(x) <= 0
            constr_eq(x) = 0
            A_ineq x <= b_ineq
            A_eq x = b_eq
            lb <= x <= ub
```
and for  CUTEst-like constraints:
```
minimize fun(x)
subject to: c_lb <= constr(x) <= c_ub
            lb <= x <= ub
```
(as I have initially proposed)

**Comment by nmayorov** [comment]
OK, I guess we need to propose/discuss specifics of `Constraints`.

Some sketch I had in my mind:

```Python

class Constraints:
   def __init__(self):
       pass  # No constraints by default.

   @classmethod
   def matlab_like(cls, ...):
       pass
   
   @classmethod
   def cutest_like(cls, ...):
       pass

   def add_box_constraints(self, lb, ub):
       pass

   def add_linear_equality(self, A, b):
       pass

   ...
```

**Comment by antonior92** [comment]
I was thinking of something entirely different. When you put like that it actually sounds like a good idea.

**Comment by josef-pkt** [comment]
I think that class interface might be too indirect to easily figure out how to specify a constraint.
I have no idea what matlab is doing, nor who or what is the cutest constraint. 

(the nice thing about dictionaries is that they are familiar if it is similar to keyword arguments, i.e. `contraints_kwds` instead of `**kwargs` )

**Comment by antonior92** [comment]
Hey @josef-pkt, I think ``matlab_like`` and ``cutest_like`` methods are only copies of the ``matlab`` and ``cutest`` interfaces for passing constraints. It should make the life easier for those familiar with those interfaces.

Nikolay's idea is actually quite good. We could use methods like:
```
   def add_box_constraints(self, lb, ub):
       pass

   def add_linear_equality(self, A, b):
       pass

   def add_nonlinear_equality(self, fun, jac, hess):
       pass

   def add_nonlinear_inequality(self, fun, jac, hess):
       pass

...
```
to define constraints in a very general way.

**Comment by dlax** [comment]
I'm missing the point of gathering several constraints into a single object. It seems to me that constraints are not inter-related in general, so a plain Python sequence is appropriate (could even be a set if we were not using `dict`s currently).
If we were to add class-based model, I'd rather consider modeling the different kind of constraints ("linear", "nonlinear", "inequality", "bounds", etc.) as individual concrete classes deriving from the same abstract base class. Though, here as well, we need to consider which methods such a class would need and if there's none, it means that a record-like data structure (i.e. a `dict` in Python) is appropriate.

**Comment by josef-pkt** [comment]
related to dlax: I was thinking that the `add_xxx` interface is not very user friendly if I want to run the estimation with different constraints imposed, i.e. drop one of the constraints to see how the result changes.

**Comment by antonior92** [comment]
@dlax your idea of having different classes modelling different classes of constraints is good. 

The advantage of using a class-based system instead of a dictionary based system is that we could include methods to converting between different types of constraints. Something like:

```
class IneqConstr:
    """Inequality constraint ``constr(x) <= 0``"""
    def __init__(self, n_vars, n_constr, constr, jac, hess):
          pass

class EqConstr:
    """Equality constraint ``constr(x) = 0``"""
    def __init__(self, n_vars, n_constr, constr, jac, hess):
        pass

class LinearEqConstr:
    def __init__(self, A, b):
         pass

    def to_nonlinear_constr(self):
        return EqConstr()


class LinearIneqConstr:
    def __init__(self, A, b):
         pass

    def to_nonlinear_constr(self):
        return IneqConstr()


class BoxConstr:
    """Box constraints of the type lb <= x <= ub"""
    def __init__(self, lb, ub):

    def to_linear_ineq_constr(self):
        return LinearIneqConstr(A, b)

    def to_nonlinear_constraint(self):
        return IneqConstr()
```
And this could be very usefull internally, since different solvers require constraints in different ways. So it would be very good to have some common implementation that could be used to do this conversions.

**Comment by shoyer** [comment]
> And this could be very usefull internally, since different solvers require constraints in different ways. So it would be very good to have some common implementation that could be used to do this conversions.

I agree that that transforming between different parameterizations seems useful, but I think this is entirely orthogonal to the public interface for `minimize`.

**Comment by mdhaber** [comment]
There seems to be a lot of debate as to whether constraints should be passed as a `dict` or as some other object. Why not a `dict`-like object that can be used either way?

**Comment by mdhaber** [comment]
Also, have we converged on working this algorithm into `minimize`, but with a backwards-compatible extension to the `constraints` parameter?

**Comment by dlax** [comment]
> Why not a dict-like object that can be used either way?

That's fine by me, though I'm not sure we should expose publicly expose
this yet.

> Also, have we converged on working this algorithm into |minimize|, but
> with a backwards-compatible extension to the |constraints| parameter?

It seems to me that the extended dict format proposed by Antonio earlier 
is backwards-compatible.

**Comment by rgommers** [comment]
>>  Why not a dict-like object that can be used either way? 

> That's fine by me, though I'm not sure we should expose publicly expose this yet.

Agree with this. In practice this means that a dict can be specified now, and the ``Constraints`` class that @nmayorov proposes can easily be added later in a backwards-compatible way (it just needs to inherit from dict or have appropriate methods like ``__index__``, ``keys``, etc.). There's no need to say "dict-like object" - ducktyping makes this automatically okay.

**Comment by antonior92** [comment]
Thank you all for the valuable suggestions! What do you think of the follow:
- Use ``minimize`` interface for both constrained and unconstrained problems.
- The constraints will be passed as a sequence of dictionaries and``dict-like objects`` in a backward compatible way.

**Comment by nmayorov** [comment]
I have questions/concerns about dictionaries specification:

1. What is the logic behind `linear` flag? If it set, then `jac` will be called only once? Can it be considered a bit tricky to grasp for people? Can we maybe introduce a separate constraint type for that? (But in this case I don't know how to make it all unified).
2. Maybe it would be better to include bound constraints as `lb`,`ub` (i.e. 2 arrays for lower and upper bounds with possible broadcasting) in `constraints` as well, and probably deprecate `bounds`. Some discussion about it #5700 Again it will create unification problems.
3. This is perhaps the biggest. Your options like `jac='2-point'` `hess` and `jas_sparsity` are assumed to work with your solver (which is great). But how are they supposed to work with existing and possible new solvers? Will they be simply ignored with a warning? What if some new solver have different options regarding constraints?

**Comment by antonior92** [comment]
A few observations about Nikolay's questions:

1. We could create a separate constraint class for linear constrains. In this case the user would not need to provide a function, but rather arrays ``A`` and ``b``. Maybe we could create a class for each constraint type:
```
class NonlinearConstraint(dict)
    def __init__(fun, jac, hess, ...)

class LinearConstraint(dict)
    def __init__(A, b)

class BoundConstraints(dict)
    def __init__(lb, ub)
```
And the user could pass a sequence of objects/dictionaries.

2. I like better to pass constraints as 2 different arrays ``lb`` and ``ub`` (as it is done in ``least_squares``), I think it is much more natural.
3. How to deal with finite-difference options (and quasi-Newton options) is a great challenge:
   - There is a function ``fprime`` used on some of the methods to estimate first derivatives. We could deprecate it or make two different ways of estimating finite diferences coexist (which does not sound like a very good idea for me).
   - If we include quasi-Newton options (``BFGS``, ``SR1`` and ``L-BFGS``) for all second-order methods, this would probably make the user very confuse about the existing methods ``BFGS`` and ``L-BFGS-B``. I don't know what would be the best way to handle this.
   - I think we should implement it in such way that it could be easily reused for the other solvers, but my priority should be to make it available to the GSoC solver I am implementing. If I have time I will make those options available to as much methods as I can. But if I am short on time I will probably just ignore it with a warning for other methods.

**Comment by mdhaber** [comment]
Hi Antonio,
1. I agree with specifying `A` and `b` for linear constraints rather than providing a function that will be called only once. When you have the chance, could you collect all your thoughts in a post like [this one](https://github.com/scipy/scipy/issues/7648#issuecomment-317730523) ("You guys are right. I am giving up too easily on..."). 
2. I don't really care whether `lb` and `ub` are separate but I agree that they should be arrays of some sort. I stuck with the list of tuples idea for `linprog` and that has been a real pain. (I should have at least converted to arrays internally but I thought someone must have had a reason not to use arrays here... still can't figure out what that was, and I wish I had converted.)
3. I understand the complexity of finite-difference/quasi-Newton options, but I don't understand the concern about what other solvers will do with the options. There are provisions made for solver-specific options (`_check_unknown_options`, `_show_options`).

**Comment by nmayorov** [comment]
Matt, we discussed with Antonio some details in our private gitter channel recently (go check :) and he will update the ideas soon.

**Comment by antonior92** [comment]
I have take some time to discuss the interface with Nikolay. What do you guys think of the following:

```python
class NonlinearConstraint:
    """Object representing nonlinear constraint

    Init Parameters
    ---------------
    fun : callable
        The function defining the constraint.
            fun(x) -> array_like, shape (m,)
    jac : {‘2-point’, ‘3-point’, ‘cs’, callable}
        If a callable is passed, it should be the Jacobian matrix of `fun`:
            jac(x) -> ndarray (or sparse matrix), shape (m, n)
        Otherwise, ‘2-point’, ‘3-point’, ‘cs’ specify the type of finite
        difference approximation used.
    hess : {‘2-point’, ‘3-point’, callable}, optional
        If a callable is passed, it should be the Hessian matrix of
        `dot(fun, v)`:
            hess(x, v) -> ndarray (or sparse matrix or LinearOperator), shape (n, n)
        Otherwise, ‘2-point’, ‘3-point’ specify the type of finite difference
        approximation used.
    constr_type : tuple
        Specifies the type of contraint. Options for this
        parameters are:
            - ("interval", lb, ub): for a constraint of the type:
                lb <= fun(x) <= ub
            - ("greater", lb): for a constraint of the type:
                fun(x) >= lb
            - ("less", ub): for a constraint of the type:
                fun(x) <= ub
            - ("equals", c): for a constraint of the type:
                fun(x) == c
            - ("greater",): for a constraint of the type:
                fun(x) >= 0
            - ("less",): for a constraint of the type:
                fun(x) <= 0
            - ("equals",): for a constraint of the type:
                fun(x) == 0
    jac_sparsity : {None, array_like, sparse matrix}, optional
        Defines the sparsity structure of the Jacobian matrix for finite
        difference estimation, its shape must be (m, n).
    """
    def __init__(self, fun, jac, hess, constr_type, jac_sparsity=None):
        pass


class LinearConstraint:
    """Object representing linear constraint.

    Init Parameters
    ---------------
    A : ndarray (or sparse matrix), shape (m, n)
        Matrix for the linear constraint.
    constr_type : tuple
        Specifies the type of contraint. Options for this
        parameters are:
            - ("interval", lb, ub): for a constraint of the type:
                lb <= A x <= ub
            - ("greater", lb): for a constraint of the type:
                A x >= lb
            - ("less", ub): for a constraint of the type:
                A x <= ub
            - ("equals", c): for a constraint of the type:
                A x == c
            - ("greater",): for a constraint of the type:
                A x >= 0
            - ("less",): for a constraint of the type:
                A x <= 0
            - ("equals",): for a constraint of the type:
                A x == 0
    """
    def __init__(self, A, constr_type):
        pass


class BoxConstraint:
    """Object representing box constraint.

    Init Parameters
    ---------------
    constr_type : tuple
        Specifies the type of contraint. Options for this
        parameters are:
            - ("interval", lb, ub): for a constraint of the type:
                lb <= x <= ub
            - ("greater", lb): for a constraint of the type:
                x >= lb
            - ("less", ub): for a constraint of the type:
                x <= ub
            - ("greater",): for a constraint of the type:
                x >= 0
            - ("less",): for a constraint of the type:
                x <= 0
    """
    def __init__(self, constr_type):
        pass
```

**Comment by mdhaber** [comment]
I like it.  Some detailed comments written for `NonlinearConstraint` (but several apply to the others):

- Specify what `n` is.  (length of x)
- Is `fun` allowed to return a 2D matrix (sparse or otherwise) of shape `(m,1`)
- What is the three point approximation of a derivative? Is the 2-point centered? (Or is 2-point one-sided and three point is centered with the weight on the center 0...?) I assume CS is complex step? 
- How do you do a 2-point approximation of a Hessian? This is a second derivative; I'm only familiar with formulas with three or more function evaluations. And I'm familiar with how these formulas are derived and I can't see how only two function evaluations can be used to approximate a Hessian.
- In `hess`, what is `v` (any arbitrary vector)? (Forgive me if I'm way off-base here, but:) Providing second order derivatives for constraints - even in optimization - is less common than for an objective function, so this may be confusing to the user (including me). You might provide a reference or more information on taking the required dot product between the 3rd-order tensor that is the second derivative of the constraints and this vector `v`. I suppose this might be interpreted as a Hessian (in a certain basis?), but it's unclear what it is the Hessian of, if we are sticking to the usual meaning of Hessian (a second partial derivatives of a scalar valued function; `fun` is not scalar-valued). So is "Hessian" the common name for this tensor-vector product?
- Specify the shape of `lb`, `ub`, and `c` (presumably `(m,)` is the typical case but see first comment about matrices). How about a scalar for broadcasting (even if there are `m` scalar constraints)? Also this is not just the constraint _type_, it is also the bounds/RHS, so consider whether the name is appropriate, or at least clarify that in the description e.g. "...type of constraint and bounds (if any).". 
- what are the other default values besides that for `jac_sparsity`?

**Comment by antonior92** [comment]
Hey Matt, thank you for the feedback! I will include a more detailed description on the actual 
documentation, this was more like a proof of concept.

A few comments bellow:
- I will specify all the shapes.
- I like the idea of including a scalar broadcasting for ``lb``, ``ub`` and ``c``.
- I will try to think a better name for ``constr_type``.
- You are right:
    hess(x, v) = sum(v[i]*H[:, :, i])
I will try to provide a more detailed documentation about it. 
- When estimating ``hess(x, v)`` using finite differences, it will always return a linear operator:
    hess(x, v) -> LinearOperator
that way there is no need to specify a sparse structure.
- The estimation of "2-point" hessian evaluate the *Jacobian* in two points. I would not recommend this option unless the analytical value for computing the *Jacobian* is actually provided (quasi-Newton methods should be used in this case).

**Comment by mdhaber** [comment]
OK that was my best guess, evaluating the (_Jacobian_ at two points). Forgot to mention it.
You might want to mention the quasi-Newton options in this documentation, right?
And what about the derivative question:
"What is the three point approximation of a derivative? Is the 2-point centered? (Or is 2-point one-sided and three point is centered with the weight on the center 0...?) I assume CS is complex step?"

**Comment by antonior92** [comment]
I just borrow this nomenclature from [least_squares](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.least_squares.html) function: ``2-point`` is foward approximation,  ``3-point`` is centered and and ``CS`` is complex step.

I will add quasi-Newton option to the documentation latter, probably something like:
```
hess : {‘2-point’, ‘3-point’, `SR1`, `BFGS`, `L-BFGS`, callable}, optional
        If a callable is passed, it should be the Hessian matrix of
        `dot(fun, v)`:
            hess(x, v) -> ndarray (or sparse matrix or LinearOperator), shape (n, n)
        Otherwise, ‘2-point’, ‘3-point’ specify the type of finite difference
        approximation used.
```

**Comment by nmayorov** [comment]
Clarification about `2-point` and `3-point`: in `least_squares` we considered bounds and made sure that evaluation points are within, which meant switching central differences to one-sided "3-point" scheme or using either forward or backward "2-point" scheme. This naming suggestion was from Evgeni Burovski. 

I think it is not strictly necessary here (hard to consider all constraints), but we may as well reuse this notation.

**Comment by felixlen** [comment]
I just saw this issue which. Something like this would be a very welcoming addition for me, especially the CUTEst like interface as I am not very happy about the current interface to the SLSQP solver.

I would like to raise to points that have not been covered yet:
- Output of results: The current output of SLSQP lacks lagrange multipliers. I have seen multiple instances of problems where SLSQP claimed convergence to points that are not optimal. Having lagrange multipliers would enable one to check KKT violation a posteriori. Also the sensitivity information provided by multipliers is sometimes very useful. Output of lagrange multiplier should reflect definition of constraints (as box constraint, linear constraint, nonlinear constraint etc).
- Seperation of Function and Derivative: In a lot of problem instances, computation of derivatives cannot be done independently of function evaluation, so splitting function and derivative causes inefficiencies and it would be very nice to have an interface for a function that yields function value and Jacobian at the same time.

**Comment by pv** [comment]
The function and derivative evaluations can be combined in minimize(), 
set jac=True

**Comment by felixlen** [comment]
@pv But only for objective function, not for constraint functions

**Comment by pv** [comment]
The lack of lagrange multipliers in SLSQP also is orthogonal to this 
issue --- they could be already to the return bunch, but this would 
require someone to do the digging in the Fortran code and find out if it 
actually returns them or not.

**Comment by pv** [comment]
Ah yes, indeed for constraints that's missing.

**Comment by antonior92** [comment]
Hi @felixlen! Thank you a lot for your input. 

I have implemented the three classes I have mentioned above, I think it is more general than the matlab-like and cutest-like interfaces I mentioned before. Furthermore, you can easily pass CUTEst-like constraints by combining a NonlinearConstraint with a BoxConstraint. About the new issues you have raised:

- Output of results: I agree with you that langrange multipliers can be really usefull. I am already returning them in the implementation presented in PR #7729. What is still missing is to return them in a way that reflects how the constraints were defined, but I inted to fix this in the next few weeks.

- About the SLSQP: by what I have understood this algorithm does not use any second order information at all, so it seems very likelly that it does not actually computes the lagrange multipliers, since the computation of those is usually done in order to be able to compute the Lagragian hessian. (Disclaimer: I have not dig into the fortran code to actually check this affirmation.)

- Separation of Function and Derivative: I agree with you that implementing function and derivatives separately is inefficient, since for many situations their computations are intrinsically related (e.g. in any algorithm that computes the derivative in a similar fashion to backward automatic differentiation). Nevertheless, it can also be inefficient to create an interface that require both of them to always be returned together (like ``minimize`` allows you to do). Something like:
```
fun, jac = fun(x)
```
That is because some times the algorithm needs to evaluate only the function (e.g. for the algorithm I have implemented it is common to call the the Jacobian only on something about 70\% of the times you need to call the actual funtion) and if every call to the function automatically calls the Jacobian you end up evaluating the Jacobian unnecessarily, what can actually be worse than trying to compute them separately.

An efficient way of implementing this would be to create an parameter that  would dictates if the Jacobian would or not be returned. Something like this:
```
fun= fun(x, return_jac=false)
fun, jac= fun(x, return_jac=true)
```
If we want we could even include the second derivative in the same computation.

**Comment by antonior92** [comment]
Some additional comment about the separation of function and derivatives: 

If you think about it the separation of function and constraint computations can also be ineficient, since these computations are related some times (e.g. nonlinear predictive control problems). The most efficient implementation would be to allow the user to pass a single callable with several parameters dictating what would be returned at each iteration. Nevertheless, this would be very not user friendly.

My point is: there is a trade off between efficiency and usability and we have to draw the line somewhere. I think the interface:
```
fun = fun(x, return_jac=false)
fun, jac = fun(x, return_jac=true)
```
(possibly including the second derivatives as well) both for function and constraints may be a good option.

**Comment by felixlen** [comment]
@antonior92 Thanks for your great work, having such interfaces and Lagrange multipliers is great. IMHO having an interface for function evaluation as you proposed in your last comment also makes a lot of sense and I would appreciate it.

As a side comment regarding SLSQP: I just dug through the fortran code. It has Lagrange multipliers (It must have! How do you check optimality otherwise?). It has two multiplier estimates (see comments from lines 151-158 in slsqp_optmz.f) in the returned workspace vector `W`:
- first `m` entries which are used in the l1-penalty function for globalization as penalty parameters and updated with the Augmented Lagrangian first order update (l.377-381), but only for constraints without simple bounds
- entries after position `m+1/2*n*(n+1)+n+1` multipliers of QP solver, for constraints and simple bounds (don't know about the order, but should be easy to find out). These should be reasonable accurate.

**Comment by pv** [comment]
@felixlen: continuing with side notes on SLSQP --- if you have your head 
wrapped around SLSQP, comments on the patch in 
https://github.com/scipy/scipy/issues/7618#issuecomment-319766332 
related to problems in its termination condition could be useful.

**Comment by antonior92** [comment]
Everything I said about the SLSQP was wrong. I should have taken a few minutes to take a look at the fortran code. :)

It does compute the Lagrange Multipliers as you have pointed out. Furthermore, it does uses second order information (By what I have understood from a quick glance at the code it uses a modified  BFGS approximation).

**Comment by antonior92** [comment]
Hi all!

I have integrated the solver to the standard ``minimize`` interface, as we have agreed upon. You can visualize the interface documentation here:
https://4938-1460385-gh.circle-artifacts.com/0/home/ubuntu/scipy/doc/build/html-scipyorg/generated/scipy.optimize.minimize.html#scipy.optimize.minimize

I called the method ``trust-constr``. This methods swiches between the two solvers I have implemented during my GSoC. I have decide to use a single method to both solvers because they are applicable in complementar situations.

The implementation is available in PR #7729. In order to keep this PR more manageable, I think it is best not to try to include the features I have implemented for the nonlinear solver available to all the other solvers at once. Hence lots of features will be available at first only for ``trust-constr`` method. Latter, in a sequence of PRs, I should include these features for other methods as well.

Some examples of features that I intend to implement in latter PRs:
- Make {'3-point', 'cs'} finite difference estimation available for estimating the Jacobian matrix for CG, BFGS, Newton-CG, L-BFGS-B, TNC, SLSQP, dogleg, trust-ncg, trust-krylov and trust-exact methods;
- Make {'2-point', '3-point', 'cs'} finite difference estimation available for estimating the Hessian matrix for CG, BFGS, Newton-CG, L-BFGS-B, TNC, SLSQP, dogleg, trust-ncg, trust-krylov and trust-exact methods;
- Make quasi-Newton approximations available for Newton-CG, dogleg, trust-ncg, trust-krylov, trust-exact and trust-constr;
- Allow Newton-CG, trust-ncg, trust-krylov to use Hessians represented by sparse matrix and Linear Operators;
- To uniformize the use of ``bounds``;
- To allow the constraint classes `BoxConstraint`, `LinearConstraint`, `NonlinearConstraint` to be used be COBYLA and SLSQP method as well;
- To allow a more complete  use of the callback function for other methods too (e.g. finish PR #7425 ).

When all these additional features have been included for the other methods we will have a better integration between the methods and a cleaner interface. Nevertheless, I believe we should left them for latter on, in order to have small and easier to review PR. Furthermore, some of these features will require me to modify well tested SciPy code and it would be better if they are commited separately and reviewed carefully in order to avoid any problems latter.

What do you guys think?

**Comment by antonior92** [comment]
>> Nevertheless, I believe we should left them for latter on, in order to have small and easier to review PR.

> For gh-7729 that ship has sailed, getting Github's pink unicorn because it's so large:) But yes, smaller is better.

:)
